pkg/cmd/psa/psa.go

```diff
@@ -0,0 +1,291 @@
+package psa
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	psapi "k8s.io/pod-security-admission/api"
+
+	"github.com/spf13/cobra"
+)
+
+// PSAOptions contains all the options and configsi for running the PSA command.
+type PSAOptions struct {
+	level      string
+	namespaces []string
+
+	client   *kubernetes.Clientset
+	warnings *warningsHandler
+}
+
+var (
+	psaExample = `
+	# Check if your namespaces could be upgraded to the restricted level.
+	%[1]s psa-check --level restricted
+
+	# Check if your namespaces could be upgraded to the restricted level.
+	%[1]s psa-check --level restricted --namespaces my-namespace,my-other-namespace
+`
+
+	empty       = struct{}{}
+	validLevels = map[string]struct{}{
+		string(psapi.LevelPrivileged): empty,
+		string(psapi.LevelBaseline):   empty,
+		string(psapi.LevelRestricted): empty,
+	}
+
+	podControllers = map[string]struct{}{
+		"Deployment":  empty,
+		"DemonSet":    empty,
+		"StatefulSet": empty,
+		"Job":         empty,
+	}
+)
+
+// NewCmdPSA creates a cobra.Command that is capable of checking namespaces for{
+// for their viability for a given PodSecurity level.
+func NewCmdPSA(parentName string, streams genericclioptions.IOStreams) *cobra.Command {
+	o := PSAOptions{}
+
+	cmd := cobra.Command{
+		Use:   "psa-check",
+		Short: "Verify namespace workloads match the namespace pod security profile",
+
+		Example: fmt.Sprintf(psaExample, parentName),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Validate(); err != nil {
+				return fmt.Errorf("validation failed: %w", err)
+			}
+			if err := o.Complete(); err != nil {
+				return fmt.Errorf("completion failed: %w", err)
+			}
+
+			return o.Run()
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.level, "level", "l", "", "The PodSecurity level to check against. The default is the audit level.")
```

Comment:
no short args.  Very few options get them becuase there are not many of them available.